src/PolyType/ReflectionProvider/ReflectionParameterShape.cs

```diff
@@ -40,8 +41,18 @@ public ReflectionParameterShape(
     ITypeShape IParameterShape.ParameterType => ParameterType;
     object? IParameterShape.Accept(TypeShapeVisitor visitor, object? state) => visitor.VisitParameter(this, state);
 
-    public Setter<TArgumentState, TParameter> GetSetter() =>
-        _setter ??= _provider.MemberAccessor.CreateConstructorArgumentStateSetter<TArgumentState, TParameter>(_ctorInfo, Position);
+    public Setter<TArgumentState, TParameter> GetSetter()
+    {
+        if (_setter is { } setter)
+        {
+            return setter;
+        }
+
+        lock (_syncObject)
+        {
+            return _setter ??= _provider.MemberAccessor.CreateConstructorArgumentStateSetter<TArgumentState, TParameter>(_ctorInfo, Position);
```

Comment:
What necessitated introduction of locks here?

Reply:
Makes sense. When writing such code I typically factor the critical path into a local method so that the getter method itself can more easily, i.e. `return _setter ?? CreateGetterHelper();`. I think we can also use `Interlocked.CompareExchange` here but the downside is potential duplicated allocations if there is contention at startup.

Reply:
Fine idea. I'll implement your `??` local function idea.